File: Native_Android_Application/app/src/test/kotlin/dev/jshstadler/carkey/DeviceHealthTest.kt
package dev.jshstadler.carkey

import org.junit.Assert.*
import org.junit.Test

class DeviceHealthTest {
    private val identity = "blekey|2|car-main|car|press,ota1,psk2,health1"
    private val snapshot = "$identity,fw=2.7.0,build=012345abcdef,idf=v6.0.1,up=90061,reset=brownout,ota=valid,heap=123456,minheap=100000,links=2,advrec=4,ghost=1"

    @Test fun parsesFirmwareSnapshotAndPreservesCapabilities() {
        val health = DeviceHealth.parse(snapshot)!!
        assertEquals("2.7.0", health.firmware)
        assertEquals(90061L, health.uptimeSeconds)
        assertEquals(4L, health.advertisingRecoveries)
        assertTrue(health.display().contains("1d 1h 1m 1s"))
        assertTrue(health.display().contains("Validated"))
        assertTrue(DeviceHealth.capabilities(snapshot).contains("psk2"))
        assertNotNull(DeviceHealth.parse("$snapshot,future=ok"))
    }

    @Test fun rejectsUnauthenticatedOldOrTruncatedResponses() {
        assertNull(DeviceHealth.parse(identity))
        assertNull(DeviceHealth.parse("blekey|2|car-main|car|press,ota1,psk2"))
        assertNull(DeviceHealth.parse(snapshot.substringBefore(",ghost=")))
        assertNull(DeviceHealth.parse(snapshot.replace("|car|", "|gate|")))
    }

    @Test fun rejectsInvalidAndAmbiguousFields() {
        listOf(snapshot.replace("up=90061", "up=-1"),
            snapshot.replace("up=90061", "up=18446744073709551615"),
            snapshot.replace("heap=123456", "heap=99999999999999999"),
            snapshot.replace("links=2", "links=4"),
            snapshot.replace("build=012345abcdef", "build=wrong"),
            snapshot.replace("minheap=100000", "minheap=200000"),
            "$snapshot,fw=wrong", snapshot.replace("fw=2.7.0", "fw=2.7.0\nwrong")
        ).forEach { assertNull(it, DeviceHealth.parse(it)) }
    }

    @Test fun pendingOtaDoesNotClaimValidation() {
        val report = DeviceHealth.parse(snapshot.replace("ota=valid", "ota=pending"))!!.display()
        assertTrue(report.contains("refresh after one minute"))
        assertFalse(report.contains("Validated"))
    }

    @Test fun parsesAndValidatesAdvertisedRadioConfiguration() {
        val data = snapshot.replace("health1,", "health1,radio1,") +
            ",fastmin=80,fastmax=160,slowmin=320,slowmax=640,idlesec=60,advmode=inactive"
        assertEquals(RadioSettings.DEFAULT, DeviceHealth.parse(data)!!.radio)
        assertEquals("inactive", DeviceHealth.parse(data)!!.advertisingMode)
        assertNull(DeviceHealth.parse(data.replace("slowmax=640", "slowmax=9999")))
        assertNull(DeviceHealth.parse(data.substringBefore(",idlesec=")))
    }
}

File: ESP32-C3_Firmware/tests/device_health_test.c
#include <assert.h>
#include <string.h>
#include "device_health.h"

int main(void) {
    char out[512];
    assert(device_health_format(out, sizeof(out), NULL) == (int)strlen(CAR_IDENTITY));
    assert(strcmp(out, CAR_IDENTITY) == 0);
    assert(strstr(out, "fw=") == NULL);
    device_health_t h = {"2.7.0", "012345abcdef", "v6.0.1", "brownout", "valid",
        UINT64_MAX, UINT32_MAX, UINT32_MAX, 3, UINT32_MAX, UINT32_MAX,
        RADIO_DEFAULTS, "recent"};
    int n = device_health_format(out, sizeof(out), &h);
    assert(n > 0 && n < (int)sizeof(out));
    assert(strstr(out, "press,ota1,psk2,health1,radio1,fw=2.7.0") != NULL);
    assert(strstr(out, "up=18446744073709551615,reset=brownout,ota=valid") != NULL);
    assert(strstr(out, "minheap=4294967295,links=3") != NULL);
    char short_out[8];
    assert(device_health_format(short_out, sizeof(short_out), &h) == n);
    assert(short_out[7] == '\0');
    return 0;
}
